### PR TITLE
#26048 add record_size_kib to aws_fsx_openzfs_volume

### LINKED ISSUE #26048: aws_fsx_openzfs: support for RecordSizeKiB
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

Add OpenZFS dataset record size configuration to tune for specific workloads.
* `record_size_kib` to `aws_fsx_openzfs_volume`
* `record_size_kib` to `root_volume_configuration` of `aws_fsx_openzfs_file_system`

### New or Affected Resource(s)

* aws_fsx_openzfs_volume
* aws_fsx_openzfs_file_system

### Potential Terraform Configuration

```hcl
resource "aws_fsx_openzfs_volume" "test" {
  name             = "testvolume"
  parent_volume_id = aws_fsx_openzfs_file_system.test.root_volume_id
  record_size_kib  = 512
}
```

### References

* #0000

### PR DESCRIPTION
**Pull request by pwillie**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Closes #26048

Output from acceptance testing:

```
$ make testacc TESTS=TestAccFSxOpenzfsVolume_recordSizeKib PKG=fsx                         
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/fsx/... -v -count 1 -parallel 20 -run='TestAccFSxOpenzfsVolume_recordSizeKib'  -timeout 180m
=== RUN   TestAccFSxOpenzfsVolume_recordSizeKib
=== PAUSE TestAccFSxOpenzfsVolume_recordSizeKib
=== CONT  TestAccFSxOpenzfsVolume_recordSizeKib
--- PASS: TestAccFSxOpenzfsVolume_recordSizeKib (1068.79s)
PASS
ok      github.com/hashicorp/terraform-provider-aws/internal/service/fsx        1070.818s
...
```

### CODE REVIEW

File: website/docs/r/fsx_openzfs_file_system.html.markdown [hunk 1]
```diff
@@ -54,6 +54,7 @@ The following arguments are supported:
 * `data_compression_type` - (Optional) - Method used to compress the data on the volume. Valid values are `LZ4`, `NONE` or `ZSTD`. Child volumes that don't specify compression option will inherit from parent volume. This option on file system applies to the root volume.
 * `nfs_exports` - (Optional) - NFS export configuration for the root volume. Exactly 1 item. See [NFS Exports](#nfs-exports) Below.
 * `read_only` - (Optional) - specifies whether the volume is read-only. Default is false.
+* `record_size_kib` - (Optional) - Specifies the record size of an OpenZFS root volume, in kibibytes (KiB). Valid values are 4, 8, 16, 32, 64, 128, 256, 512, or 1024 KiB. The default is 128 KiB.
```
Comment (bschaatsbergen) [inline website/docs/r/fsx_openzfs_file_system.html.markdown:57]:
```suggestion
* `record_size_kib` - (Optional) - Specifies the record size of an OpenZFS root volume, in kibibytes (KiB). Valid values are `4`, `8`, `16`, `32`, `64`, `128`, `256`, `512`, or `1024` KiB. The default is `128` KiB.
```

File: website/docs/r/fsx_openzfs_volume.html.markdown [hunk 2]
```diff
@@ -31,6 +31,7 @@ The following arguments are supported:
 * `data_compression_type` - (Optional) Method used to compress the data on the volume. Valid values are `NONE` or `ZSTD`. Child volumes that don't specify compression option will inherit from parent volume. This option on file system applies to the root volume.
 * `nfs_exports` - (Optional) NFS export configuration for the root volume. Exactly 1 item. See [NFS Exports](#nfs-exports) Below.
 * `read_only` - (Optional) specifies whether the volume is read-only. Default is false.
+* `record_size_kib` - (Optional) The record size of an OpenZFS volume, in kibibytes (KiB). Valid values are 4, 8, 16, 32, 64, 128, 256, 512, or 1024 KiB. The default is 128 KiB.
```
Comment (bschaatsbergen) [inline website/docs/r/fsx_openzfs_volume.html.markdown:34]:
```suggestion
* `record_size_kib` - (Optional) The record size of an OpenZFS volume, in kibibytes (KiB). Valid values are `4`, `8`, `16`, `32`, `64`, `128`, `256`, `512`, or `1024` KiB. The default is `128` KiB.
```

**Comment by bschaatsbergen** [comment]
Hi @pwillie - thanks for raising this PR, could you please add a changelog entry?

File: website/docs/r/fsx_openzfs_file_system.html.markdown (hunk 1 above)
Comment (pwillie) [reply website/docs/r/fsx_openzfs_file_system.html.markdown]: Added suggested change.

File: website/docs/r/fsx_openzfs_volume.html.markdown (hunk 2 above)
Comment (pwillie) [reply website/docs/r/fsx_openzfs_volume.html.markdown]: Added suggested change.

**Comment by pwillie** [comment]
Hi @bschaatsbergen - changelog added (although think I referenced the issue rather than pull request.  Is that ok?)

**Comment by bschaatsbergen** [comment]
> Hi @bschaatsbergen - changelog added (although think I referenced the issue rather than pull request. Is that ok?)

Preferably we reference the pull request

**Comment by pwillie** [comment]
Good to know.  Updated.

**Comment by bschaatsbergen** [comment]
LGMT :rocket: 

The tests that failed here - already fail on main
`bruno@pop-os ~/G/f/pwillie-terraform-provider-aws (fsx-openzfs-recordsize)> make testacc TESTS=TestAccFSxOpenzfsVolume_ PKG=fsx                                                                                                 (base) 
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/fsx/... -v -count 1 -parallel 20 -run='TestAccFSxOpenzfsVolume_'  -timeout 180m
=== RUN   TestAccFSxOpenzfsVolume_basic
=== PAUSE TestAccFSxOpenzfsVolume_basic
=== RUN   TestAccFSxOpenzfsVolume_parentVolume
=== PAUSE TestAccFSxOpenzfsVolume_parentVolume
=== RUN   TestAccFSxOpenzfsVolume_tags
=== PAUSE TestAccFSxOpenzfsVolume_tags
=== RUN   TestAccFSxOpenzfsVolume_copyTags
=== PAUSE TestAccFSxOpenzfsVolume_copyTags
=== RUN   TestAccFSxOpenzfsVolume_name
=== PAUSE TestAccFSxOpenzfsVolume_name
=== RUN   TestAccFSxOpenzfsVolume_dataCompressionType
=== PAUSE TestAccFSxOpenzfsVolume_dataCompressionType
=== RUN   TestAccFSxOpenzfsVolume_readOnly
=== PAUSE TestAccFSxOpenzfsVolume_readOnly
=== RUN   TestAccFSxOpenzfsVolume_recordSizeKib
=== PAUSE TestAccFSxOpenzfsVolume_recordSizeKib
=== RUN   TestAccFSxOpenzfsVolume_storageCapacity
=== PAUSE TestAccFSxOpenzfsVolume_storageCapacity
=== RUN   TestAccFSxOpenzfsVolume_nfsExports
=== PAUSE TestAccFSxOpenzfsVolume_nfsExports
=== RUN   TestAccFSxOpenzfsVolume_userAndGroupQuotas
=== PAUSE TestAccFSxOpenzfsVolume_userAndGroupQuotas
=== CONT  TestAccFSxOpenzfsVolume_basic
=== CONT  TestAccFSxOpenzfsVolume_readOnly
=== CONT  TestAccFSxOpenzfsVolume_copyTags
=== CONT  TestAccFSxOpenzfsVolume_userAndGroupQuotas
=== CONT  TestAccFSxOpenzfsVolume_parentVolume
=== CONT  TestAccFSxOpenzfsVolume_tags
=== CONT  TestAccFSxOpenzfsVolume_dataCompressionType
=== CONT  TestAccFSxOpenzfsVolume_nfsExports
=== CONT  TestAccFSxOpenzfsVolume_name
=== CONT  TestAccFSxOpenzfsVolume_storageCapacity
=== CONT  TestAccFSxOpenzfsVolume_recordSizeKib
=== CONT  TestAccFSxOpenzfsVolume_basic
    openzfs_volume_test.go:24: Step 1/2 error: Check failed: Check 9/15 error: aws_fsx_openzfs_volume.test: Attribute 'nfs_exports.0.client_configurations.0.options.#' expected "1", got "2"
=== CONT  TestAccFSxOpenzfsVolume_userAndGroupQuotas
    openzfs_volume_test.go:412: Step 1/3 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # aws_fsx_openzfs_volume.test will be updated in-place
          ~ resource "aws_fsx_openzfs_volume" "test" {
                id                               = "fsvol-092edc483b8e6de8e"
                name                             = "tf-acc-test-5224169345004751284"
                tags                             = {}
                # (10 unchanged attributes hidden)
        
              - user_and_group_quotas {
                  - id                         = 0 -> null
                  - storage_capacity_quota_gib = 0 -> null
                  - type                       = "GROUP" -> null
                }
              - user_and_group_quotas {
                  - id                         = 0 -> null
                  - storage_capacity_quota_gib = 0 -> null
                  - type                       = "USER" -> null
                }
        
                # (2 unchanged blocks hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccFSxOpenzfsVolume_userAndGroupQuotas (975.51s)
--- FAIL: TestAccFSxOpenzfsVolume_basic (993.59s)
--- PASS: TestAccFSxOpenzfsVolume_tags (1029.17s)
--- PASS: TestAccFSxOpenzfsVolume_parentVolume (1088.77s)
--- PASS: TestAccFSxOpenzfsVolume_storageCapacity (1092.46s)
--- PASS: TestAccFSxOpenzfsVolume_copyTags (1099.85s)
--- PASS: TestAccFSxOpenzfsVolume_nfsExports (1132.36s)
--- PASS: TestAccFSxOpenzfsVolume_name (1133.99s)
--- PASS: TestAccFSxOpenzfsVolume_dataCompressionType (1154.14s)
--- PASS: TestAccFSxOpenzfsVolume_recordSizeKib (1154.81s)
--- PASS: TestAccFSxOpenzfsVolume_readOnly (1154.94s)
FAIL
FAIL    github.com/hashicorp/terraform-provider-aws/internal/service/fsx        1155.001s
FAIL
make: *** [GNUmakefile:52: testacc] Error 1`

**Review by bschaatsbergen — APPROVED** [review]

**Comment by pwillie** [comment]
Thanks for the review @bschaatsbergen

**Comment by ewbankkit** [comment]
```console
% make testacc TESTARGS='-run=TestAccFSxOpenzfsVolume_' PKG=fsx ACCTEST_PARALLELISM=2
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./internal/service/fsx/... -v -count 1 -parallel 2  -run=TestAccFSxOpenzfsVolume_ -timeout 180m
=== RUN   TestAccFSxOpenzfsVolume_basic
=== PAUSE TestAccFSxOpenzfsVolume_basic
=== RUN   TestAccFSxOpenzfsVolume_parentVolume
=== PAUSE TestAccFSxOpenzfsVolume_parentVolume
=== RUN   TestAccFSxOpenzfsVolume_tags
=== PAUSE TestAccFSxOpenzfsVolume_tags
=== RUN   TestAccFSxOpenzfsVolume_copyTags
=== PAUSE TestAccFSxOpenzfsVolume_copyTags
=== RUN   TestAccFSxOpenzfsVolume_name
=== PAUSE TestAccFSxOpenzfsVolume_name
=== RUN   TestAccFSxOpenzfsVolume_dataCompressionType
=== PAUSE TestAccFSxOpenzfsVolume_dataCompressionType
=== RUN   TestAccFSxOpenzfsVolume_readOnly
=== PAUSE TestAccFSxOpenzfsVolume_readOnly
=== RUN   TestAccFSxOpenzfsVolume_recordSizeKib
=== PAUSE TestAccFSxOpenzfsVolume_recordSizeKib
=== RUN   TestAccFSxOpenzfsVolume_storageCapacity
=== PAUSE TestAccFSxOpenzfsVolume_storageCapacity
=== RUN   TestAccFSxOpenzfsVolume_nfsExports
=== PAUSE TestAccFSxOpenzfsVolume_nfsExports
=== RUN   TestAccFSxOpenzfsVolume_userAndGroupQuotas
=== PAUSE TestAccFSxOpenzfsVolume_userAndGroupQuotas
=== CONT  TestAccFSxOpenzfsVolume_basic
=== CONT  TestAccFSxOpenzfsVolume_readOnly
--- PASS: TestAccFSxOpenzfsVolume_basic (933.18s)
=== CONT  TestAccFSxOpenzfsVolume_copyTags
--- PASS: TestAccFSxOpenzfsVolume_readOnly (1079.16s)
=== CONT  TestAccFSxOpenzfsVolume_dataCompressionType
--- PASS: TestAccFSxOpenzfsVolume_copyTags (1066.28s)
=== CONT  TestAccFSxOpenzfsVolume_name
--- PASS: TestAccFSxOpenzfsVolume_dataCompressionType (1067.53s)
=== CONT  TestAccFSxOpenzfsVolume_nfsExports
=== CONT  TestAccFSxOpenzfsVolume_userAndGroupQuotas
--- PASS: TestAccFSxOpenzfsVolume_name (1108.95s)
--- PASS: TestAccFSxOpenzfsVolume_nfsExports (1081.44s)
=== CONT  TestAccFSxOpenzfsVolume_tags
=== CONT  TestAccFSxOpenzfsVolume_userAndGroupQuotas
    openzfs_volume_test.go:412: Step 1/3 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
        stdout
        
        
        Terraform used the selected providers to generate the following execution
        plan. Resource actions are indicated with the following symbols:
          ~ update in-place
        
        Terraform will perform the following actions:
        
          # aws_fsx_openzfs_volume.test will be updated in-place
          ~ resource "aws_fsx_openzfs_volume" "test" {
                id                               = "fsvol-0dfe5f622b4dbaee2"
                name                             = "tf-acc-test-6341542772351364948"
                tags                             = {}
                # (10 unchanged attributes hidden)
        
        
              - user_and_group_quotas {
                  - id                         = 0 -> null
                  - storage_capacity_quota_gib = 0 -> null
                  - type                       = "GROUP" -> null
                }
              - user_and_group_quotas {
                  - id                         = 0 -> null
                  - storage_capacity_quota_gib = 0 -> null
                  - type                       = "USER" -> null
                }
                # (2 unchanged blocks hidden)
            }
        
        Plan: 0 to add, 1 to change, 0 to destroy.
--- PASS: TestAccFSxOpenzfsVolume_tags (922.15s)
=== CONT  TestAccFSxOpenzfsVolume_storageCapacity
--- FAIL: TestAccFSxOpenzfsVolume_userAndGroupQuotas (979.88s)
=== CONT  TestAccFSxOpenzfsVolume_recordSizeKib
--- PASS: TestAccFSxOpenzfsVolume_recordSizeKib (1117.72s)
=== CONT  TestAccFSxOpenzfsVolume_parentVolume
--- PASS: TestAccFSxOpenzfsVolume_storageCapacity (1158.71s)
--- PASS: TestAccFSxOpenzfsVolume_parentVolume (1026.34s)
FAIL
FAIL	github.com/hashicorp/terraform-provider-aws/internal/service/fsx	6317.256s
FAIL
make: *** [testacc] Error 1
```

Failures are unrelated to this change and occur in nightly CI:

```
=== RUN   TestAccFSxOpenzfsVolume_userAndGroupQuotas
=== PAUSE TestAccFSxOpenzfsVolume_userAndGroupQuotas
=== CONT  TestAccFSxOpenzfsVolume_userAndGroupQuotas
openzfs_volume_test.go:376: Step 1/3 error: After applying this test step and performing a `terraform refresh`, the plan was not empty.
stdout
Terraform used the selected providers to generate the following execution
plan. Resource actions are indicated with the following symbols:
~ update in-place
Terraform will perform the following actions:
# aws_fsx_openzfs_volume.test will be updated in-place
~ resource "aws_fsx_openzfs_volume" "test" {
id                               = "fsvol-0156f0af32bc6bc04"
name                             = "tf-acc-test-5453947432521315878"
tags                             = {}
# (9 unchanged attributes hidden)
- user_and_group_quotas {
- id                         = 0 -> null
- storage_capacity_quota_gib = 0 -> null
- type                       = "GROUP" -> null
}
- user_and_group_quotas {
- id                         = 0 -> null
- storage_capacity_quota_gib = 0 -> null
- type                       = "USER" -> null
}
# (2 unchanged blocks hidden)
}
Plan: 0 to add, 1 to change, 0 to destroy.
--- FAIL: TestAccFSxOpenzfsVolume_userAndGroupQuotas (1018.14s)
```

**Review by ewbankkit — APPROVED** [review]
LGTM :rocket:.

**Comment by ewbankkit** [comment]
@pwillie Thanks for the contribution :tada: :clap:.

**Pull request merged**